omniledger/ChildTransactions.md

```diff
@@ -0,0 +1,120 @@
+
+down the communication tree, and then every
+node will respond with a type
+
+```
+message PollTxResponse {
+  repeated Transaction Txs = 1;
+}
+```
+
+The transactions are combined on the subleader nodes.
+
+However, before sending the `PollTxResponse` message, the conodes must check that
+the state of omniledger does include the transactions in the latest block
+given by the id in `PollTxRequest`. If the state is not
+up-to-date, then the nodes must do an update to ensure it is. Then, the nodes
+verify `Transaction.Signature` to make sure that all transaction in their queue are
+valid. The valid transactions are marked as "Verified" and the bad transactions
+are dropped and a message is printed to the audit log. Finally, the transactions
+with the "Verified" flag are sent to the leader in the `PollTxResponse` message.
+These transactions are marked as "Submitted". The `PollTxResponse` message should
+not be larger than B bytes.
+
+Upon receiving all the `PollTxResponse` message, the leader does the following:
+- remove duplicates
+- verify signature
+- sort the transactions in a random but deterministic way
+- go through the list of transactions, and for each transaction mark if it
+applies correctly to the state updated with previous valid transactions
+
+Then the leader creates the block and then sends it to the conodes to cosign, e.g.
+
+```
+message BlockProposal {
+  Data Data = 1;
+}
+```
+
+The conodes run the same checks and need to make sure that the transactions are
+in the same state as marked by the leader. If this is the case, they sign the
+hash of the proposed block.
+
+The new block, with a collective signature, is propagated back to all nodes.
+Then every node updates their queue and removes the transactions that are in the
+new block. For the transactions that were not added to the new block, they need
+to be moved to the front of the queue and marked as "New" because the state
+of omniledger may have changed and the old transactions may become invalid. All the
+"Verified" transactions must also be changed back to "New".
+
+### Additional blocks
+
+What we described above is how to generate a single block, how do we run it
+multiple times? A simple solution would be for the leader to send a
+`PollTxRequest` after every new block is generated. However, it results in a lot
+of wasted messages if there are very little or no transactions. We can attempt
+to implement the simplest technique first and then try to optimise it later. For
+example, a slightly better version would be to add some delay when the blocks
+are getting smaller. But this is only a heuristic because the leader does not
+know how many transactions are in the queues of the non-root nodes.
+
+Another option would be that each node sends _one_ message to the leader if it
+has not-included transactions. This could happen when:
+- the queue has been empty and a first transaction comes in
+- a new block has been accepted, but not all transaction of the queue are in
```

Comment:
How about piggybacking an advisory of "I have some txns you should come pick up" on the replies to request to sign a new block?

Reply:
Sounds like a possible optimisation, yes.